File: 2069-walking-robot-simulation-ii/2069-walking-robot-simulation-ii.java
class Robot {

    int m, n;
    int x, y;
    char direction;
    int len;

    public Robot(int width, int height) {
        this.m = height;
        this.n = width;
        this.x = 0;
        this.y = 0;
        this.direction = 'E';
        this.len = 0;
    }
    
    public void step(int num) {
        if (num == 0) return;
        int totalLen = 2 * (m + n) - 4;
        int rem = (len + num) % totalLen;
        len = rem;
        if (rem == 0) {
            x = 0;
            y = 0;
            direction = 'S';
        } else if (rem > 2 * n + m - 3) {
            x = 0;
            y = totalLen - rem;
            direction = 'S';
        } else if (rem > n + m - 2) {
            x = 2*n+m-3-rem;
            y = m-1;
            direction = 'W';
        } else if (rem > n - 1) {
            x = n-1;
            y = rem-n+1;
            direction = 'N';
        } else {
            x = rem;
            y = 0;
            direction = 'E';
        }
    }
    
    public int[] getPos() {
        return new int[]{x, y};
    }
    
    public String getDir() {
        switch (direction) {
            case 'N': return "North";
            case 'S': return "South";
            case 'E': return "East";
            default: return "West";
        }
    }
}

File: 2069-walking-robot-simulation-ii/2069-walking-robot-simulation-ii.cpp
class Robot {
public:
    Robot(int width, int height){
        directions["East"] = {1,0};
        directions["North"] = {0,1};
        directions["South"] = {0,-1};
        directions["West"] = {-1,0};
        
        row = col = 0;
        W = width;
        H = height;
        paramter = 2*(H+W-2);
        order = {"East", "North", "West", "South"};
        curr_dir = 0;
    }
    
    void step(int num) {
        if(row == 0 && col == 0){
            num = num%paramter;
            if(num == 0) curr_dir = 3;
        }
        row += num * directions[order[curr_dir]].first;
        col += num * directions[order[curr_dir]].second;
        bool exceeded = true;

        if(row >= W){
            num = row - (W-1);
            row = W-1;
        }
        else if(col >= H){
            num = col - (H-1);
            col = H-1;
        }
        else if (row < 0){
            num = abs(row);
            row = 0;
        }
        else if(col < 0){
            num = abs(col);
            col = 0;
        }
        else{
           exceeded = false; 
        }


        if(exceeded){
            curr_dir = (curr_dir + 1)%4;
            step(num);
        }
    }
    
    vector<int> getPos() {
        return {row, col};
    }
    
    string getDir() {
        return order[curr_dir];
    }
    unordered_map<string, pair<int,int>> directions;
    int curr_dir;
    int row, col;
    int W, H;
    int paramter;
    vector<string> order;
};